language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -166,71 +266,137 @@ instantiateEvaluate con ty fun
 -- If succesful, proceed with either this same term or with the result of the computation
 -- depending on whether 'BuiltinName' is saturated or not.
 applyEvaluate
-    :: VarEnv -> VarEnv -> Context -> Plain Value -> Plain Value -> CekM EvaluationResultDef
+    :: VarEnv
+    -> VarEnv
+    -> Context
+    -> WithMemory Value
+    -> WithMemory Value
+    -> CekM (Plain Term)
 applyEvaluate funVarEnv argVarEnv con (LamAbs _ name _ body) arg =
     withVarEnv (extendVarEnv name arg argVarEnv funVarEnv) $ computeCek con body
-applyEvaluate funVarEnv _         con fun                    arg =
-    let term = Apply () fun arg in
+applyEvaluate funVarEnv _ con fun arg =
+    let term = Apply 1 fun arg
+    in -- TODO
         case termAsPrimIterApp term of
-            Nothing                       ->
-                throwError $ MachineException NonPrimitiveApplicationMachineError term
+            Nothing -> throwError $ CekInternalError $ MachineException
+                NonPrimitiveApplicationMachineError
+                (void term)
             Just (IterApp headName spine) -> do
-                constAppResult <- applyStagedBuiltinName headName spine
+                constAppResult <- applyStagedBuiltinName arg headName spine
                 withVarEnv funVarEnv $ case constAppResult of
                     ConstAppSuccess res -> computeCek con res
-                    ConstAppFailure     -> pure EvaluationFailure
-                    ConstAppStuck       -> returnCek con term
-                    ConstAppError   err ->
-                        throwError $ MachineException (ConstAppMachineError err) term
+                    ConstAppFailure ->
+                        throwError $ CekUserError $ CekEvaluationFailure
+                    ConstAppStuck -> returnCek con term
+                    ConstAppError err ->
+                        throwError $ CekInternalError $ MachineException
+                            (ConstAppMachineError err)
+                            (void term)
 
-evaluateInCekM :: EvaluateConstApp (Either CekMachineException) a -> CekM (ConstAppResult a)
-evaluateInCekM a =
-    ReaderT $ \cekEnv ->
-        let eval means' = evaluateCekIn $ cekEnv & cekEnvMeans %~ mappend means'
-            in runEvaluateConstApp eval a
+evaluateInCekM :: EvaluateConstApp CekM a -> CekM (ConstAppResult a)
+evaluateInCekM a = do
+    let eval means' term =
+            local (over cekEnvMeans (<> means'))
+                $ fmap EvaluationSuccess
+                $ computeCek [] (withMemory term) -- TODO 'withMemory' has a cost, should be avoided
```

Comment:
The constant application machinery doesn't handle annotations currently, but it should. @kwxm did you get annotations working?

Reply:
Still stands.